### PR TITLE
Add class and instance-level Client config defaults

### LINKED ISSUE #113: Add support for boolean config variables
Use case: `permissive_ssl`.

Currently, every non-null value is true.

It would be nice to support: `on`, `off`, `true`, `false`, `0`, `1`.

### LINKED ISSUE #425: Ability to "hint" config options from user-code without hard override of the user environment config
With the current 3 layers to provide config options (like `solver` or `endpoint`):
1) config file (all options)
2) environment variables (limited set of options)
3) keyword arguments to Client's constructor and `from_config` (all options)

it is impossible to "suggest" default option values (e.g. solver), but allow for user to override it on any of the 3 layers. Possible use cases:
- user wants to set a soft default in code (layer 3), but allow override from environment
- 3rd party library has a notion of preferred defaults, but wants to allow an easy override for its users

Proposal: implement a mechanism to provide config options on layer zero -- a per-Client instance default value for each config option, provided on construction.

For example, adding `defaults` argument to `Client` constructor and `from_config`, which would `.update` the `Client` class-level defaults for all accepted config options. Additional benefits of this are: consolidation of the defaults, clearly documenting all accepted config options in one place. It can be nicely combined with #357.

### PR DESCRIPTION
**Pull request by randomir**
Closes #425.
Closes #113.

### CODE REVIEW

File: dwave/cloud/client.py [hunk 1]
```diff
@@ -81,60 +82,86 @@
...
-        polling_timeout (float, default=None):
-            Problem status polling timeout (in seconds), after which polling is aborted.
+        polling_timeout (float, optional):
+            Problem status polling timeout (in seconds), after which polling is
+            aborted.
 
         connection_close (bool, default=False):
-            Force HTTP(S) connection close after each request.
+            Force HTTP(S) connection close after each request. Use to prevent
+            intermediate network equipment closing idle connections.
```
Comment (JoelPasvolsky) [inline dwave/cloud/client.py:127]: "Use to prevent..." --> "Set to True to ..."

File: dwave/cloud/client.py [hunk 2]
```diff
@@ -81,60 +82,86 @@
...
+
+        headers (dict/str, optional):
+            Newline-separated additional HTTP headers to include with each
+            API request, or a dictionary of (key, value) pairs.
+
+        client_cert (str, optional):
+            Client certificate file path.
+
+        client_cert_key (str, optional):
+            Client certificate key file path.
```
Comment (JoelPasvolsky) [inline dwave/cloud/client.py:137]: Five nouns in a row is tough to parse. "File path to client's certificate key" would be easier on the eye.

File: dwave/cloud/client.py [hunk 3]
```diff
@@ -81,60 +82,86 @@
...
+
+        poll_backoff_min (float, default=0.05):
+            Problem status polling exponential back-off base period, in seconds.
 
-        headers (dict/str):
-            Additional HTTP headers.
+        poll_backoff_max (float, default=60):
+            Problem status polling exponential back-off max period, in seconds.
```
Comment (JoelPasvolsky) [inline dwave/cloud/client.py:143]: Okay, compared to a row of seven nouns, five wasn't too bad. "Maximum exponential back-off period, in seconds, for polling problem status."

File: dwave/cloud/client.py [hunk 4]
```diff
@@ -81,60 +82,86 @@
...
 
-    Other Parameters:
-        Unrecognized keys (str):
-            All unrecognized keys are passed through to the appropriate client class constructor
-            as string keyword arguments.
+        defaults (dict, optional):
+            Client instance-level default, overriding class-level
```
Comment (JoelPasvolsky) [inline dwave/cloud/client.py:146]: Do users need those "level"s? Something like, "Defaults for the client instance that override the class :attr:`.DEFAULTS`."

File: dwave/cloud/client.py [hunk 5]
```diff
@@ -144,13 +171,36 @@ class Client(object):
...
+        'solver': None,
+        'proxy': None,
+        'permissive_ssl': False,
+        'request_timeout': 60,
+        'polling_timeout': None,
+        'connection_close': False,
+        'headers': None,
+        'client_cert': None,
+        'client_cert_key': None,
+        # Poll back-off schedule defaults [sec]
```
Comment (JoelPasvolsky) [inline dwave/cloud/client.py:199]: Why do these two get their own section?

File: dwave/cloud/client.py [hunk 6]
```diff
@@ -81,60 +82,86 @@
...
 
         request_timeout (float, default=60):
-            Connect and read timeout (in seconds) for all requests to the D-Wave API.
+            Connect and read timeout (in seconds) for all requests to the
+            D-Wave API.
 
-        polling_timeout (float, default=None):
-            Problem status polling timeout (in seconds), after which polling is aborted.
+        polling_timeout (float, optional):
+            Problem status polling timeout (in seconds), after which polling is
```
Comment (JoelPasvolsky) [inline dwave/cloud/client.py:122]: "timeout, in seconds, after"

File: dwave/cloud/client.py [hunk 7]
```diff
@@ -81,60 +82,86 @@
...
+
+        proxy (str, optional):
             Proxy URL to be used for accessing the D-Wave API.
 
         permissive_ssl (bool, default=False):
             Disables SSL verification.
 
         request_timeout (float, default=60):
-            Connect and read timeout (in seconds) for all requests to the D-Wave API.
+            Connect and read timeout (in seconds) for all requests to the
```
Comment (JoelPasvolsky) [inline dwave/cloud/client.py:118]: timeout, in seconds, for all

File: dwave/cloud/client.py [hunk 8]
```diff
@@ -195,95 +236,32 @@ def from_config(cls, config_file=None, profile=None, client=None,
         1. Values specified as keyword arguments in :func:`from_config()`
         2. Values specified as environment variables
         3. Values specified in the configuration file
+        4. Values specified as :class:`.Client` instance defaults
+        5. Values specified in :class:`.Client` class :attr:`.Client.DEFAULTS`.
```
Comment (JoelPasvolsky) [inline dwave/cloud/client.py:240]: Only item 5 has a period

File: dwave/cloud/client.py [hunk 9]
```diff
@@ -195,95 +236,32 @@ def from_config(cls, config_file=None, profile=None, client=None,
...
-        an environment variable.
+        Configuration-file format and environment variables are described in
+        :mod:`dwave.cloud.config`.
 
-        Environment variables: ``DWAVE_CONFIG_FILE``, ``DWAVE_PROFILE``,
-        ``DWAVE_API_CLIENT``, ``DWAVE_API_ENDPOINT``, ``DWAVE_API_TOKEN``,
-        ``DWAVE_API_SOLVER``, ``DWAVE_API_PROXY``, ``DWAVE_API_HEADERS``.
-
-        Environment variables are described in :mod:`dwave.cloud.config`.
+        Configuration file and environment variables loading mechanism is
```
Comment (JoelPasvolsky) [inline dwave/cloud/client.py:245]: is --> are
"loading mechanism for environment variables"

File: dwave/cloud/config.py [hunk 10]
```diff
@@ -33,6 +33,9 @@
 1. Values specified as keyword arguments
 2. Values specified as environment variables
 3. Values specified in the configuration file
+4. Values specified in :class:`~dwave.cloud.client.Client` instance ``defaults``
+5. Values specified in :class:`~dwave.cloud.client.Client` class
+   :attr:`~dwave.cloud.client.Client.DEFAULTS`.
```
Comment (JoelPasvolsky) [inline dwave/cloud/config.py:39]: Only item 5 has period 
(You're consistent)

File: dwave/cloud/config.py [hunk 11]
```diff
@@ -621,23 +666,27 @@ def get_default_config():
...
     order (with 1 the highest ranked):
 
     1. Values specified as keyword arguments in :func:`load_config()`. These values replace
        values read from a configuration file, and therefore must be **strings**, including float
-       values for timeouts, boolean flags (tested for "truthiness"), and solver feature
+       values for timeouts, boolean flags, and solver feature
        constraints (a dictionary encoded as JSON).
     2. Values specified as environment variables.
     3. Values specified in the configuration file.
+    4. Values specified as :class:`~dwave.cloud.client.Client` instance defaults
```
Comment (JoelPasvolsky) [inline dwave/cloud/config.py:681]: period

**Review by JoelPasvolsky — APPROVED** [review]
Nice

File: docs/reference/resources.rst [hunk 12]
```diff
@@ -27,31 +27,40 @@ that it later explicitly closes.
...
-   client.Client.get_solvers
-   client.Client.is_solver_handled
-   client.Client.close
+   Client.from_config
+   Client.solvers
+   Client.retrieve_answer
+   Client.get_solver
+   Client.get_solvers
+   Client.is_solver_handled
+   Client.close
```
[the remark this answers is not in the document]
Comment (randomir) [reply docs/reference/resources.rst:62]: There are a few more missing all over the place. Don't like them sorted yet. We'll need a bigger refactor.

File: dwave/cloud/client.py [hunk 13]
```diff
@@ -81,60 +82,86 @@
...
+
+            Defined via dictionary of solver feature constraints
+            (see :meth:`~dwave.cloud.client.Client.get_solvers`).
+            For backward compatibility, a solver name, as a string, is also
+            accepted and converted to ``{"name": <solver name>}``.
 
-        proxy (str):
+            If undefined, :meth:`~dwave.cloud.client.Client.get_solver` uses a
+            solver definition from environment variables, a configuration file,
+            or falls back to the first available online solver.
```
[the remark this answers is not in the document]
Comment (randomir) [reply dwave/cloud/client.py:109]: I moved this from `from_config`, I think, but yes, TMI. Dropping from here.

**Review by arcondello — APPROVED** [review]
LGTM

File: dwave/cloud/client.py (hunk 5 above)
Comment (randomir) [reply dwave/cloud/client.py:199]: Starting a trend of options that will follow. :smiley:

**Pull request merged**